cpp/src/table/row_operators.cu

```diff
@@ -344,39 +345,227 @@ namespace row {
 
 namespace lexicographic {
 
+namespace {
+
+/**
+ * @brief Transform a lists-of-structs column into lists-of-integers column.
+ *
+ * For lists-of-structs column at any nested level, the child structs column will be replaced by an
+ * integer column of its ranks generated using `cudf::rank()`.
+ *
+ * If the input column is not lists-of-structs, or does not contain lists-of-structs at any nested
+ * level, the input will be passed through.
+ *
+ * @param input The input column to transform
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return A pair of new column_view representing the transformed input and the generated rank
+ *         (integer) column which needs to be kept alive.
+ */
+std::pair<column_view, std::unique_ptr<column>> transform_lists_of_structs(
+  column_view const& input, rmm::cuda_stream_view stream)
+{
+  auto const make_transformed_input = [&](auto const& new_child) {
+    return column_view{data_type{type_id::LIST},
+                       input.size(),
+                       nullptr,
+                       input.null_mask(),
+                       input.null_count(),
+                       input.offset(),
+                       {input.child(lists_column_view::offsets_column_index), new_child}};
+  };
+
+  if (input.type().id() == type_id::LIST) {
+    // Should not use sliced child because we reuse the input's offset value and offsets child.
+    auto const child = input.child(lists_column_view::child_column_index);
+
+    if (child.type().id() == type_id::STRUCT) {
+      // Dense ranks can accurately reflect the order of structs: structs compared equal will have
+      // the same rank values.
+      // However, first ranks are computed faster and are good enough for ordering them. Structs
+      // compared equal always have consecutive rank values (in stable order) thus they are still
+      // sorted correctly by their ranks.
+      auto ranks             = cudf::detail::rank(child,
+                                      rank_method::FIRST,
+                                      order::ASCENDING,
+                                      null_policy::EXCLUDE,
+                                      null_order::BEFORE,
+                                      false,
+                                      stream,
+                                      rmm::mr::get_current_device_resource());
+      auto transformed_input = make_transformed_input(ranks->view());
+      return {std::move(transformed_input), std::move(ranks)};
+    } else if (child.type().id() == type_id::LIST) {
+      auto [new_child, ranks_cols] = transform_lists_of_structs(child, stream);
+      if (ranks_cols) {
+        auto transformed_input = make_transformed_input(new_child);
+        return {transformed_input, std::move(ranks_cols)};
+      }
+    }
+  } else if (input.type().id() == type_id::STRUCT) {
+    CUDF_UNREACHABLE("Structs columns should be flattened before calling this function.");
```

Comment:
Can we not get struct columns at this point? After all, we are only flattening instances of `struct<list<...>>`

Reply:
If there is any `struct<list<>>` column, the entire input table will be flattened. This helps simplifying other subsequent operations a lot. Otherwise, we have to:
 * Make correction of the column order and null order for both non-flattened and flattened columns, and
 * Maintains a vector of `unique_ptr<flattened_table>` to store different results from flattening columns.
 * Have to consider `list<struct<>>` while transforming lists of structs in this function, and return a list of columns (`vector<unique_ptr<column>>`) instead of just one column because we may have structs of multiple lists columns having structs child.

Reply:
Ah, I see the `std::any_of` in `has_nested_structs_of_lists`. Can we not move the `has_nested_structs_of_lists` check [here](https://github.com/rapidsai/cudf/blob/96e28b8629a129adab8f011b92eae87b93958673/cpp/src/structs/utilities.cpp#L171) instead? That way, we can always call `flatten_nested_structs` without any if-else. Any struct columns that have no lists in them will go in the `else` portion of the code I linked above

Reply:
That way we still flatten the entire table, but reduce the bookkeeping overhead by flattening only structs of lists. No correction to column/null order will be needed, and we'll be able to get away with a single `unique_ptr<flattened_table>`

Reply:
Unfortunatley no. That utility function needs to **always** flatten structs columns, and is used in other places such as in join.

Reply:
Even if we re-implement that utility function (so we can solve the first point I listed above), we still have to deal with 2 remaining points.

Reply:
Yes, for now. But if you think this strategy is viable then I can revisit this as a follow-on once all my comparator update PRs go in and I remove explicit flattenings from the whole code-base. Then, we'll only have implicit flattenings left through the new comparator.

I don't think we'll face point 2 we'll still flatten the whole table, but the actual struct flattenings will be limited to struct of lists. Other struct columns will be stored as is.

I will put some more thought for point 3. You are right that this strategy breaks for `struct<list<struct<>>>` because even after the conversion we are left with `struct<list<size_type>>`. I was hoping `struct<list<size_type>>` would work out-of-the-box with the new comparator.

But yes, for now we will keep the implementation as you have written it since the alternative is not viable yet.

Reply:
> adding a parameter allowing to choose if we want to flatten any structs columns, or flatten only structs-have-lists columns

Yes!

Here's what we can do for this PR:
1. Move `has_nested_structs_of_lists` to `structs/utilities.cpp`
2. Add a default parameter to `flatten_nested_columns`. Default option is to flatten only structs of lists, `join` and other affected algorithms can use non-default option to flatten all structs

Once I finish the comparator update, I remove the new parameter entirely and we only flatten struct of lists from that point onwards.

However, if you'd rather not worry about it in this PR then I'll take it up as a follow on after my comparator updates

Reply:
Got it, thanks. Then let's do it in the follow-up PR instead, so we can minimize code change in this one.